### PR TITLE
Span tutorial

### LINKED ISSUE #812: Add a short tutorial for using spans in your model
This could be a more explanatory tutorial, or a shorter how-to, whichever makes the most sense.  It'd be nice to have this to point to in our release notes that mention the span representations.

### PR DESCRIPTION
**Pull request by DeNeutoy**
Fixes #812 First pass at a tutorial - i'm not great at these, comments welcome

### CODE REVIEW

File: tutorials/how_to/span_representations.md [hunk 1]
```diff
@@ -0,0 +1,180 @@
...
+
+## Extracting Span Representations from a text sequence
+
+In many cases, you will want to extract spans from vector representations of sentences.
+In order to do this in AllenNLP, you will need to use a `SpanExtractor`. Broadly, a 
+`SpanExtractor` takes a sequence tensor of shape `(batch_size, sentence_length, embedding_size)`
+and some indices of shape `(batch_size, num_spans, 2)` and return the relevant indices into
+the sequence as a tensor of shape `(batch_size, num_spans, encoded_size)`.
+
+The simplest `SpanExtractor` is the `EndpointSpanExtractor`, which epresents spans as a
```
[the remark this answers is not in the document]
Comment (matt-gardner) [reply tutorials/how_to/span_representations.md]: And making `EndpointSpanExtractor` a link to the file (or the documentation for the file) would be good.

File: tutorials/how_to/span_representations.md [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+                                      [3, 4]]]))
+
+# We concatenated the representations for the start and end of
+# the span, so the embedded span size is 2 * embedding_size.
+# shape (batch_size, num_spans, 2 * embedding_size).
+span_representations = extractor(sequence_tensor, indices)
+assert list(span_representations.size()) == [2, 2, 14]
+```
+
+There are other types of Span Extractors - for instance, the `SelfAttentiveSpanExtractor`,
```
Comment (matt-gardner) [inline tutorials/how_to/span_representations.md]: Similarly making the `SelfAttentiveSpanExtractor` a link would be good.

File: tutorials/how_to/span_representations.md [hunk 3]
```diff
@@ -0,0 +1,180 @@
...
+which computes span representations by generating an unnormalized attention score for each
+word in the sentence. Spans representations are then computed with respect to these
+scores by normalising the attention scores for words inside the span.
+
+
+## Scoring and Pruning Spans
+
+Span based representations have been effective for modelling/approximating structured
```
Comment (matt-gardner) [inline tutorials/how_to/span_representations.md]: `s/Span based/Span-based/`

File: tutorials/how_to/span_representations.md [hunk 4]
```diff
@@ -0,0 +1,180 @@
...
+prediction problems - however, many models which leverage this type of representation
+also involve some kind of _span enumeration_ (i.e considering all possible spans in a 
+sentence/document). For a given sentence of length n, there are n^2 spans. In itself,
```
Comment (matt-gardner) [inline tutorials/how_to/span_representations.md]: Making this (and below) a [superscript](https://stackoverflow.com/a/15156585/1467533) would be nice.

File: tutorials/how_to/span_representations.md [hunk 5]
```diff
@@ -0,0 +1,180 @@
...
+### Heuristically prune spans in your DatasetReader.
+
+We have added a utility method for enumerating all spans in a sentence, but excluding
+those which fulfil some condition based on the input text or any Spacy `Token` attribute.
+For instance, for co-reference, all spans which are mentions (spans which are 
+co-referent with _something_) never start or end with punctuation, because of the way
+the Onotonotes 5.0 dataset was created. This means that we can exclude any span which
+starts or ends with punctuation using a very simple python function:
+
+```
```
Comment (matt-gardner) [inline tutorials/how_to/span_representations.md]: Mark this one as python, too.

File: tutorials/how_to/span_representations.md [hunk 6]
```diff
@@ -0,0 +1,180 @@
...
+## Generating `SpanFields` from text in a `DatasetReader`
+
+`SpanFields` are a type of `Field` in AllenNLP which take a start index, an end index
+and a `SequenceField` which the indices refer to. Once a batch of `SpanFields` has been
+converted to a tensor, we will have a matrix of shape (batch_size, 2), where the last
+dimension contains the start and end indices passed in to the SpanField constructor.
+However, for many models, you'll want to represent _many_ spans for a single batch
+element - the way to do this is to use a `ListField[SpanFields]`, which will create
+a tensor of shape (batch_size, num_spans, 2) once indexed.
+
```
[the remark this answers is not in the document]
Comment (matt-gardner) [reply tutorials/how_to/span_representations.md]: Ah, I see that you did that below.  My first thought would be to put it here, but if you think it flows better the way it is, that's fine.

File: tutorials/how_to/span_representations.md [hunk 7]
```diff
@@ -0,0 +1,180 @@
...
+                                    filter_function=no_prefixed_punctuation)
+
+# 'spans' won't include (2, 4) or (3, 4) as these have
+# punctuation as their last element. Note that these spans
+# have inclusive start and end indices!
+assert spans == [(0, 1), (0, 2), (1, 2), (1, 3), (2, 3)]
+```
+
+There are other helpful functions in `allennlp.data.dataset_readers.dataset_utils.span_utils`,
+such as a function to convert between BIO labelings and span based representations.
```
Comment (matt-gardner) [inline tutorials/how_to/span_representations.md]: `s/span based/span-based/

File: tutorials/how_to/span_representations.md [hunk 8]
```diff
@@ -0,0 +1,180 @@
...
+
+### Heuristically prune spans in your DatasetReader.
+
+We have added a utility method for enumerating all spans in a sentence, but excluding
+those which fulfil some condition based on the input text or any Spacy `Token` attribute.
+For instance, for co-reference, all spans which are mentions (spans which are 
+co-referent with _something_) never start or end with punctuation, because of the way
+the Onotonotes 5.0 dataset was created. This means that we can exclude any span which
+starts or ends with punctuation using a very simple python function:
+
```
Comment (matt-gardner) [inline tutorials/how_to/span_representations.md]: You might mention in here also that you probably don't want cross-sentence spans, which cuts down on complexity, also.

File: tutorials/how_to/span_representations.md [hunk 9]
```diff
@@ -0,0 +1,180 @@
...
+There are other helpful functions in `allennlp.data.dataset_readers.dataset_utils.span_utils`,
+such as a function to convert between BIO labelings and span based representations.
+
+### Use a SpanPruner
+
+It's not always possible to prune spans before they enter your model. AllenNLP contains
+a `SpanPruner`, which allows you to prune spans based on a parameterised function which
+is trained end-to-end with the rest of your model.
+
+```
```
Comment (matt-gardner) [inline tutorials/how_to/span_representations.md]: Mark as python.

File: tutorials/how_to/span_representations.md [hunk 10]
```diff
@@ -0,0 +1,180 @@
...
+
+spans = span_utils.enumerate_spans(sentence,
+                                    max_span_width=3,
+                                    min_span_width=2,
+                                    filter_function=no_prefixed_punctuation)
+
+# 'spans' won't include (2, 4) or (3, 4) as these have
+# punctuation as their last element. Note that these spans
+# have inclusive start and end indices!
+assert spans == [(0, 1), (0, 2), (1, 2), (1, 3), (2, 3)]
```
Comment (matt-gardner) [inline tutorials/how_to/span_representations.md]: One argument for putting this above when you talk about `DatasetReaders`: you could actually give a full example where you use this function and create a `ListField[SpanField]`.

File: tutorials/how_to/span_representations.md [hunk 11]
```diff
@@ -0,0 +1,180 @@
...
+pruner = SpanPruner(scorer=linear_scorer)
+
+# Here we'll create some spans from a random tensor of shape
+# (batch_size, num_spans, embedding_size). Typically this would
+# be the output of a SpanExtractor applied to some encoded representation
+# of a sentence, such as the output of an LSTM, or word embeddings.
+spans = Variable(torch.randn([3, 4, 5]))
+mask = Variable(torch.ones([3, 4]))
+
+# Ok, there's quite a bit to unpack here.
```
Comment (matt-gardner) [inline tutorials/how_to/span_representations.md]: I'd probably remove the "Ok,".

File: tutorials/how_to/span_representations.md [hunk 12]
```diff
@@ -0,0 +1,180 @@
...
+# See below for a full explaination.
+pruned_embeddings, pruned_mask, pruned_indices, pruned_scores = pruner(spans, mask, num_spans_to_keep=3)
+```
+
+1. First, we've got our `pruned_embeddings`.
```
Comment (matt-gardner) [inline tutorials/how_to/span_representations.md]: It's not immediately obvious that you're continuing here where the comment left off.  You might put a line like "A `SpanPruner` has four return values:" before this list.

File: tutorials/how_to/span_representations.md [hunk 13]
```diff
@@ -0,0 +1,180 @@
...
+mask = Variable(torch.ones([3, 4]))
+
+# Ok, there's quite a bit to unpack here.
+# See below for a full explaination.
+pruned_embeddings, pruned_mask, pruned_indices, pruned_scores = pruner(spans, mask, num_spans_to_keep=3)
+```
+
+1. First, we've got our `pruned_embeddings`. 
+These are of shape `(batch_size, num_spans_to_keep, embedding_size)`
+The spans we kept corespond to the topk with respect to the parameterised
```
Comment (matt-gardner) [inline tutorials/how_to/span_representations.md]: `s/corespond/correspond/`

Comment (matt-gardner) [reply tutorials/how_to/span_representations.md]: `s/topk/top k/`

File: tutorials/how_to/span_representations.md [hunk 14]
```diff
@@ -0,0 +1,180 @@
...
+span scorer. The other spans just get discarded, and your eventual loss
+function for your model won't be a function of the discarded spans!
+
+2. Secondly, we've got the `pruned_mask`, which has shape `(batch_size, num_spans_to_keep)`.
+In 99% of cases, this will be all ones. However, if you have masked spans in a 
+batch element, and you request that the `SpanPruner` keeps more than the number
+of non-masked spans, there will be some masked elements in the returned spans.
+
+3. Thirdly, we have the `pruned_indices` which has shape `(batch_size, num_spans_to_keep)`
+The indices of the top-k scoring spans into the original ``spans`` tensor.
```
Comment (matt-gardner) [inline tutorials/how_to/span_representations.md]: The hyphen in "top-k" makes me think the bracketing is "(top-k) (scoring spans)", instead of "(top k scoring) spans".  I'd also probably use "in" instead of "into" here.

File: tutorials/how_to/span_representations.md [hunk 15]
```diff
@@ -0,0 +1,180 @@
+
+Using Span Representations in AllenNLP
+--------------------------------------
+_Note that this tutorial goes through some quite advanced
+usage of AllenNLP - you may want to familiarise yourself with the repository
```
Comment (joelgrus) [inline tutorials/how_to/span_representations.md]: familiariZe 🇺🇸

File: tutorials/how_to/span_representations.md (hunk 3 above)
Comment (joelgrus) [reply tutorials/how_to/span_representations.md]: modeLing 🇺🇸

File: tutorials/how_to/span_representations.md [hunk 16]
```diff
@@ -0,0 +1,180 @@
...
+
+tokenizer = SpacyWordSplitter(pos_tags=True)
+sentence = tokenizer.split_words("This is a sentence.")
+
+def no_prefixed_punctuation(tokens: List[Token]) -> bool:
+    # Only include spans which don't start or end with punctuation.
+    return tokens[0].pos_ != "PUNCT" and tokens[-1].pos_ != "PUNCT"
+
+spans = span_utils.enumerate_spans(sentence,
+                                    max_span_width=3,
```
Comment (joelgrus) [inline tutorials/how_to/span_representations.md]: indentation is not quite right here

File: tutorials/how_to/span_representations.md [hunk 17]
```diff
@@ -0,0 +1,180 @@
...
+assert spans == [(0, 1), (0, 2), (1, 2), (1, 3), (2, 3)]
+```
+
+There are other helpful functions in `allennlp.data.dataset_readers.dataset_utils.span_utils`,
+such as a function to convert between BIO labelings and span based representations.
+
+### Use a SpanPruner
+
+It's not always possible to prune spans before they enter your model. AllenNLP contains
+a `SpanPruner`, which allows you to prune spans based on a parameterised function which
```
Comment (joelgrus) [inline tutorials/how_to/span_representations.md]: parametrized 🇺🇸

File: tutorials/how_to/span_representations.md (hunk 13 above)
Comment (joelgrus) [reply tutorials/how_to/span_representations.md]: parametrized 🇺🇸

File: tutorials/how_to/span_representations.md [hunk 18]
```diff
@@ -0,0 +1,180 @@
...
+
+# Here we'll create some spans from a random tensor of shape
+# (batch_size, num_spans, embedding_size). Typically this would
+# be the output of a SpanExtractor applied to some encoded representation
+# of a sentence, such as the output of an LSTM, or word embeddings.
+spans = Variable(torch.randn([3, 4, 5]))
+mask = Variable(torch.ones([3, 4]))
+
+# Ok, there's quite a bit to unpack here.
+# See below for a full explaination.
```
Comment (joelgrus) [inline tutorials/how_to/span_representations.md]: explanation

File: tutorials/how_to/span_representations.md (hunk 14 above)
Comment (joelgrus) [reply tutorials/how_to/span_representations.md]: I feel like you need some sort of conjunction between these two sentences

File: tutorials/how_to/span_representations.md [hunk 19]
```diff
@@ -0,0 +1,180 @@
+
+Using Span Representations in AllenNLP
+--------------------------------------
+_Note that this tutorial goes through some quite advanced
+usage of AllenNLP - you may want to familiarise yourself with the repository
+before you go through this Span Representation Tutorial._
+
+Many state of the art Deep NLP models use representations of spans,
+rather than representations of words, as the basic building block for
+models. From AllenNLP 0.4, Span Representations are extremely easy to
```
[the remark this answers is not in the document]
Comment (joelgrus) [reply tutorials/how_to/span_representations.md]: it might be worth saying a little bit about how you used to represent spans and why this approach is better
